src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -85,26 +84,67 @@ public Task RunAsync(ActionRunStage stage)
                 actionRunner.Stage = stage;
                 actionRunner.Condition = aStep.Condition;
 
-                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, location, Environment);
-
+                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, Environment);
                 InitializeScope(step);
 
-                location++;
+                // Add all steps to the Composite StepsRunner
+                // Follows similar logic to how JobRunner invokes the StepsRunner for job steps!
+                compositeSteps.Add(step);
             }
 
-            // Create a step that handles all the composite action steps' outputs
-            Pipelines.ActionStep cleanOutputsStep = new Pipelines.ActionStep();
-            cleanOutputsStep.ContextName = ExecutionContext.ContextName;
-            // Use the same reference type as our composite steps.
-            cleanOutputsStep.Reference = Action;
+            try
+            {
+                // This is where we run each step.
+                await RunStepsAsync(compositeSteps, ExecutionContext);
+
+                // This is where we set the outputs.
+                // All steps have the same scope for their ExecutionContext/
+                // So we can use one of them so that we can get the right "outputs" attribute.
+                if (compositeSteps.Count > 0)
+                {
+                    // TODO: Find better solution. 
+                    HandleOutput(compositeSteps[0].ExecutionContext);
+                }
+            }
+            catch (Exception ex)
+            {
+                // Composite StepRunner should never throw exception out.
+                Trace.Error($"Caught exception from composite steps {nameof(CompositeActionHandler)}: {ex}");
+                ExecutionContext.Error(ex);
+                ExecutionContext.Result = TaskResult.Failed;
+            }
+        }
+
+        private void CompleteStep(IStep step, TaskResult? result = null, string resultCode = null)
+        {
+            var executionContext = step.ExecutionContext;
+
+            executionContext.Complete(result, resultCode: resultCode);
```

Comment:
not sure whether should ever call `ExecutionContext.Complete()` for steps inside a composite action. `ExecutionContext.Complete()` will update the timeline record.

Reply:
Yeah, now that I think about it, we don't really need to call Complete() in each step's execution context. 

It doesn't really do anything since we'll eventually call Complete() in the StepsRunner for the whole composite action step in the job. 

I'll take it out.

Reply:
Yeah, this is confusing, I'll remove it. I was just using that for failing a step. I'll consolidate the code and just put the parts we need for failing a step.